File: crates/agent/src/phase.rs
//! Phases as data (ADR-010 Option C). A phase is a named configuration of
//! the paper and nearly nothing else (§9); keeping it data keeps the
//! phase-cut question open as a config question and lets the forge one day
//! propose phase changes through the same gates as any module.

use serde::{Deserialize, Serialize};

use context::Budget;
use kernel::{PhaseId, ToolId};

/// Which capabilities a phase exposes. `None` is structural, not refused:
/// `render` receives it and emits no tool schema at all — Plan and Verify
/// cannot act even if the model asks (ADR-010).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum ToolScope {
    None,
    /// Everything THIS agent's `agent.md` gave it. The per-agent decision is
    /// the frontmatter's (Python: `tools:` is the toolkit); the phase's job is
    /// only to say whether this phase may act at all.
    All,
    Only(Vec<ToolId>),
}

/// The exact reply shape a phase demands — parsed, never trusted. One
/// variant per §9 contract; a new contract is a new variant plus a parser
/// arm, surfaced as a design change rather than smuggled in via prose.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ResponseContract {
    /// Ordered steps + success criteria (Plan).
    PlanSteps,
    /// Exactly one tool envelope (Work).
    ToolEnvelope,
    /// pass / fail / retry / replan + reason (Verify).
    Verdict,
    /// Prose to the user — the cheap exit every graph must have
    /// (RESEARCH phase-cut).
    Answer,
}

/// Verify's four legal judgments (ADR-010). An enum so an illegal fifth
/// judgment is a parse error the machine handles, not a string the loop
/// improvises on.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum Verdict {
    Pass,
    Fail,
    Retry,
    Replan,
}

/// What a parsed reply can signal about where to go next. Matched against a
/// phase's `exits`; anything unmatched is an illegal transition the machine
/// repairs or fails deterministically (ADR-010).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ExitCondition {
    PlanProduced,
    ToolResultReceived,
    VerdictPass,
    VerdictFail,
    VerdictRetry,
    VerdictReplan,
    AnswerProduced,
}

/// Where an exit leads. `Answer` is a terminal, not a phase — modeling it
/// here keeps `PhaseId` honest (there is no "Answer document" to configure).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum PhaseExit {
    To(PhaseId),
    Answer,
}

/// One phase, fully described (ADR-010's record, field for field). Static
/// data in v1, validated in `cargo test`; the interpreter in `step` stays
/// thin — a phase needing bespoke code is a design smell to surface.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct PhaseConfig {
    pub phase: PhaseId,
    pub contract: ResponseContract,
    pub tools: ToolScope,
    pub budget: Budget,
    /// The ONLY legal next phases; everything else is handled by the
    /// machine, never by prose.
    pub exits: Vec<(ExitCondition, PhaseExit)>,
}

/// WHAT ONE WORKING TURN'S PAPER MAY COST. **PROVISIONAL (§17)** — 4096 was
/// too small for the agent that actually ships, and it was failing silently.
///
/// Measured 2026-08-23, `main` with its shipped peer `critic` loaded, in
/// `work`, asked "what is in this folder?": the paper wanted 4174 tokens before
/// a conversation existed, so the ladder pointered `## history`, pointered
/// `## space` and ELIDED `## observations` — on every single turn. The agent's
/// own prose tells it to read `## observations`; the budget was deleting the
/// block. `crates/agent/tests/prompt.rs` fails on exactly that now, by name.
///
/// 8192 and not more: the standing paper is ~4.2k (soul 2187 + affordances
/// ~1450 + environment 370), `compact_at: 8` means the window must hold eight
/// turns beside it, and doubling leaves ~4k for them while keeping the whole
/// prompt inside a 16k context — the floor for the local models this page is
/// pointed at. A number chosen against a MEASURED model context would be
/// better than one chosen against this arithmetic; that measurement is the
/// owner's to take, which is why this is marked provisional and not settled.
const WORK_BUDGET: Budget = Budget { max_tokens: 8192 };

/// The v1 phase set: Work/Verify default with Plan-on-demand and Answer as
/// the cheap exit (RESEARCH phase-cut softening of §9's symmetric cut —
/// changing this back is a config edit, which is the point of Option C).
pub fn v1_phases() -> Vec<PhaseConfig> {
    vec![
        // Work: one conversational turn that may act. The contract is
        // ToolEnvelope — prose is still a legal reply and ends the turn (the
        // cheap exit), tool calls run and come back as observations. `All`
        // means "this agent's own toolbox", which `agent.md`'s `tools:` key
        // decides; `render` builds the affordances section from exactly that
        // toolbox, so no ad-hoc prompt string can name a tool the agent was
        // not given (ADR-010, I13). A hardcoded list here was the bug the
        // Agents card exposed: the file said `tools:` and nothing read it.
        PhaseConfig {
            phase: PhaseId::Work,
            contract: ResponseContract::ToolEnvelope,
            tools: ToolScope::All,
            budget: WORK_BUDGET,
            exits: vec![
                (ExitCondition::ToolResultReceived, PhaseExit::To(PhaseId::Work)),
                (ExitCondition::AnswerProduced, PhaseExit::Answer),
            ],
        },
        // Verify: configured but unreachable until Work emits tool effects.
        PhaseConfig {
            phase: PhaseId::Verify,
            contract: ResponseContract::Verdict,
            tools: ToolScope::None,
            budget: Budget { max_tokens: 2048 },
            exits: vec![
                (ExitCondition::VerdictPass, PhaseExit::Answer),
                (ExitCondition::VerdictRetry, PhaseExit::To(PhaseId::Work)),
                (ExitCondition::VerdictReplan, PhaseExit::To(PhaseId::Plan)),
            ],
        },
    ]
}

File: packages/agent/src/phase.js
/**
 * The phase vocabulary, and the tool grant a phase makes.
 *
 * IN RUST THIS WAS A MACHINE. `v1_phases()` returned a table looked up by
 * `AgentState.phase`, and each row carried an `exits` table saying where a
 * reply could lead. The table had ONE row; `phase` had no assignment anywhere
 * outside the constructor; `exits` had a write and no reader. So the lookup
 * could only ever return Work and the exits could only ever be ignored. The
 * table, the id field, the exit conditions and the exit targets are all gone,
 * and the single row survives as a constant: a machine that never had a second
 * state is not a machine, and porting one would be porting the intention.
 *
 * What survives is the part that is READ every call — the contract a reply is
 * parsed against, and the scope that decides whether this phase may act at all.
 * @module
 */

/** @typedef {import('@harness/kernel').ToolId} ToolId */
/** @typedef {import('@harness/kernel').PhaseId} PhaseId */

/**
 * Which of an agent's tools a phase exposes. `none` is structural, not a
 * refusal at call time: the granted set is what the affordances section is
 * built from, so a phase that may not act cannot even NAME a tool to the model.
 * @typedef {{kind: 'none'} | {kind: 'all'} | {kind: 'only', tools: ToolId[]}} ToolScope
 */

/** @type {ToolScope} */
export const NO_TOOLS = Object.freeze({ kind: 'none' })

/**
 * Everything THIS agent's `agent.md` gave it. The per-agent decision belongs to
 * the frontmatter's `tools:` key; a phase says only whether it may act.
 * @type {ToolScope}
 */
export const ALL_TOOLS = Object.freeze({ kind: 'all' })

/** @param {readonly ToolId[]} tools @returns {ToolScope} */
export function onlyTools(tools) {
  return { kind: 'only', tools: [...tools] }
}

/**
 * The scope applied to one agent's toolbox — the ONLY narrowing there is.
 *
 * Generic over the element rather than over a `Toolbox` type because the whole
 * of the operation is a filter by name, and pinning it to the descriptor shape
 * would make the phase know what a tool is when all it knows is which names it
 * allows. Order is the TOOLBOX's, never the scope's: the model reads these as
 * lines, and the agent file decided their order.
 * @template {{name: string}} T
 * @param {ToolScope} scope
 * @param {readonly T[]} tools
 * @returns {T[]}
 */
export function grant(scope, tools) {
  if (scope.kind === 'none') return []
  if (scope.kind === 'all') return [...tools]
  return tools.filter((tool) => scope.tools.includes(tool.name))
}

/**
 * The exact reply shape a phase demands back — parsed, never trusted.
 *
 * TWO, DOWN FROM FOUR. `PlanSteps` and `Verdict` were deleted from the Rust in
 * 2026-08-23 with zero construction sites between them; the stage machine does
 * the planning and the judging now, as prose in files a person can edit.
 * @typedef {'tool_envelope' | 'answer'} ResponseContract
 */

/** @type {readonly ResponseContract[]} */
export const RESPONSE_CONTRACTS = /** @type {const} */ (['tool_envelope', 'answer'])

/** @typedef {{maxTokens: number}} Budget */

/**
 * WHAT ONE WORKING TURN'S PAPER MAY COST. **PROVISIONAL** — 4096 was too small
 * for the agent that actually ships, and it failed silently.
 *
 * Measured 2026-08-23 on the Rust build, `main` with its peer `critic` loaded,
 * asked "what is in this folder?": the paper wanted 4174 tokens before a
 * conversation existed, so the ladder pointered `## history`, pointered
 * `## space` and ELIDED `## observations` — on every turn. The agent's own
 * prose tells it to read `## observations`; the budget was deleting the block.
 *
 * 8192 and not more: the standing paper is ~4.2k, `compactAt` means the window
 * must hold turns beside it, and doubling leaves ~4k for them while keeping the
 * whole prompt inside a 16k context — the floor for the local models this page
 * is pointed at. A number chosen against a MEASURED model context would be
 * better than one chosen against this arithmetic; that measurement is the
 * owner's, which is why this is provisional and not settled.
 * @type {Budget}
 */
export const WORK_BUDGET = Object.freeze({ maxTokens: 8192 })

/**
 * @typedef {{phase: PhaseId, contract: ResponseContract, tools: ToolScope, budget: Budget}} PhaseConfig
 */

/**
 * ONE WORKING TURN, and the only configuration there is.
 *
 * The contract is `tool_envelope`, which does not mean prose is illegal: prose
 * is the cheap exit every graph must have, and it ends the turn. `all` means
 * "this agent's own toolbox", because a hardcoded list here was the bug the
 * Agents card exposed — the file said `tools:` and nothing read it.
 * @type {PhaseConfig}
 */
export const WORK = Object.freeze({
  phase: /** @type {PhaseId} */ ('work'),
  contract: /** @type {ResponseContract} */ ('tool_envelope'),
  tools: ALL_TOOLS,
  budget: WORK_BUDGET,
})
